docs/design/proposals/metro-dr.md

```diff
@@ -0,0 +1,70 @@
+cluster will be moved to another kubernetes cluster (expected that
+namespace name and the pvc name remains same) as a single ceph cluster is used
+by all kubernetes cluster the cephcsi running on the cluster where
+failover/fallback is happening should be able to identify the underlying
+RBD/CephFS volume and reattach to the PV.
+
+## Architecture
+
+![metro-dr](metro-dr.png)
+
+### Design
+
+Currently, ceph-csi generates its unique ID for volume and stores the
+corresponding PV name and the unique ID mapping, It creates the backend volume
+with the unique ID and returns the encoded value which contains all the
+required information for other operations. For metro DR, the same volume needs
+to be used in the secondary Kubernetes cluster. In the secondary cluster, the
+cephcsi cannot identify the volume corresponding to the PVC. Based on the
+assumption that the Kubernetes namespace and the PVC name remain the same when
+admin is doing the failover/fallback cephcsi can use the PVC name and namespace
+name to identify the RBD image. For that cephcsi needs to make a few changes.
+
+### Pre-req
+
+* The PVC name and namespace need to be passed to the CSI driver as part of the
+  volume create request.
+* A new storage-class parameter to indicate that image name need to be derived
+  from PVC name and namespace name.
+* An unique ID in the cephcsi ceph cluster configmap
+
+In the case of the primary cluster as the PVC name and the namespace name
+should be sent in volume create request, the cephcsi will generate omap data
+and creates its request name (PV name) and unique ID mapping in omap
+`csi.volume.<uniqueID>.PVName` : `volume UUID` (this is required to identify
+the RBD image/cephFS subvolume belongs to the PV) and creates an RBD image with
+the name `csi-vol-<pvcname>-<namespacename>-<UUID>`.  When doing the failover,
+cephcsi in the secondary cluster will check the volume is already present, If
+the image is present cephcsi will add its request name(PV Name) mapping to the
+UUID (`csi.volume.<uniqueID>.PVName` : `volume UUID`). As ID passed in
+configmap is unique across all the cephcsi instances accessing the ceph cluster
+will help cephcsi to identify the volume.
+
+Currently, we are making use of watchers in node stage request to make sure RWO
+PVC is mounted on a single node at a given point of time. We need to change the
+watchers logic in the node stage request as when we enable the RBD mirroring on
+an image a watcher will be added on an RBD image.
+
+In case of the PVC is Retained and recreated with the same name in the same
+namespace as the request name is unique cephcsi will not return the underlying
+storage. But if the PVC is Retained on one cluster and PVC gets created with
+the same name in the same namespace in another cluster will get underlying
+storage (it won't create a new volume (the support for the need to be added
+later by marking RBD/CephFS storage for not to be reused))
+
```

Comment:
Can we also add in some pseudo-code as we think this will behave, say:

```
Pseudo code:
  - CreateVolume
    - CheckVolExists
    - if PVCNaming:
      - List and check if image exists, and is still unbound to a request in this cluster, if so bind it and update common image OMap
    - Else:
      - Create a new volume
  - DeleteVolume
    - Determine image OMap, and store it
    - Delete image
    - Delete image OMap
    - Recursively delete CSI OMap keys from RequestName in image OMap
```